src/index.ts

```diff
@@ -315,428 +317,457 @@ export async function onFormSubmit(
+      const businessLogicStart = performance.now();
+      runStatus = await runBusinessLogics(txn.get, action);
+      const businessLogicEnd = performance.now();
+      const businessLogicLogicResult: LogicResult = {
+        name: "runBusinessLogics",
         status: "finished",
         documents: [],
-        execTime: securityFnEnd - securityFnStart,
+        execTime: businessLogicEnd - businessLogicStart,
       };
-      logicResults.push(securityLogicResult);
-      if (securityResult.status === "rejected") {
-        console.log(`Security check failed: ${securityResult.message}`);
-        await formRef.update({"@status": "security-error", "@messages": securityResult.message});
+      logicResults.push(businessLogicLogicResult);
+
+      await actionRef.set(action);
+      async function saveLogicResults() {
+        for (let i = 0; i < runStatus.logicResults.length; i++) {
+          const {documents, ...logicResult} = runStatus.logicResults[i];
+          const logicResultsRef = actionRef.collection("logicResults")
+            .doc(`${actionRef.id}-${i}`);
+          await logicResultsRef.set(logicResult);
+          const documentsRef = logicResultsRef.collection("documents");
+          for (let j = 0; j < documents.length; j++) {
+            await documentsRef.doc(`${logicResultsRef.id}-${j}`).set(documents[j]);
+          }
+        }
+      }
+      await saveLogicResults();
+
+      if (runStatus.status === "cancel-then-retry") {
+        await formRef.update({"@status": "cancelled", "@messages": "cancel-then-retry received " +
+              "from business logic"});
+        const end = performance.now();
+        const execTime = end - start;
+        const onFormSubmitLogicResult: LogicResult = {
+          name: "onFormSubmit",
+          status: "finished",
+          documents: [],
+          execTime,
+        };
+        logicResults.push(onFormSubmitLogicResult);
+        await indexUtilsMockable.createMetricExecution(logicResults);
         return;
       }
-    }
 
-    // Check for delay
-    console.info("Checking for delay");
-    const delay = form["@delay"];
-    if (delay) {
-      const delayStart = performance.now();
-      const cancelled = await delayFormSubmissionAndCheckIfCancelled(delay, formRef);
-      const delayEnd = performance.now();
-      const delayLogicResult: LogicResult = {
-        name: "delayFormSubmission",
+      const distributeLogicResultsStart = performance.now();
+
+      let errorMessage= "";
+      function updateErrorMessage() {
+        const errorLogicResults = runStatus.logicResults.filter((result) => result.status === "error");
+        if (errorLogicResults.length > 0) {
+          errorMessage = errorMessage + errorLogicResults.map((result) => result.message).join("\n");
+        }
+      }
+      updateErrorMessage();
+
+      await distributeFnTransactional(txn, runStatus.logicResults);
+      await formRef.update({"@status": "finished"});
```

Comment:
This update in status should happen after non-transactional logic results are distributed.

Reply:
From this line to end of this transactional block should happen after the distribution of non transactional logic result